LiteDB/Engine/Disk/MemoryCache.cs

```diff
@@ -401,28 +435,186 @@ private void Extend()
         public int Clear()
         {
             var counter = 0;
+            var now = DateTime.UtcNow;
+            var overCap = false;
 
             ENSURE(this.PagesInUse == 0, "must have no pages in use when call Clear() cache");
 
-            foreach (var page in _readable.Values)
+            // Thread-safe enumeration of ConcurrentDictionary
+            foreach (var kv in _readable)
             {
-                page.Position = long.MaxValue;
-                page.Origin = FileOrigin.None;
+                if (_readable.TryRemove(kv.Key, out var page))
+                {
+                    // Reset page controls
+                    page.Position = long.MaxValue;
+                    page.Origin = FileOrigin.None;
+
+                    // Mark as "freed at" for idle policy
+                    page.Timestamp = now.Ticks;
 
-                _free.Enqueue(page);
+                    _free.Enqueue(page);
+                    counter++;
 
-                counter++;
+                    // Check if we exceed free pages limit
+                    if (_free.Count > _maxFreePages)
```

Comment:
``_free.Count`` enumerates the whole collection. Try to use a cheaper volatile int and ``Interlocked.`` on it.